opengovsg/plumber
# PLU-91: [SGID-LOGIN-2] Implement SGID login button

## Problem
We want to support SGID login as an alternative to OTP.

## Solution
As opposed to the usual SGID approach of creating routes on the back end, we will implement SGID using front end routes + GraphQL instead (to avoid us mixing and matching classic REST vs GraphQL where possible).

### High Level Approach
_See #202 for full details._

1. Create a new "login with SGID" GraphQL mutation, which logs in users with a single work email, or return the list of work emails to the front end.
2. Create a new "login with selected SGID" mutation, which logs multi-hatted users in after they have selected their email.
3. Create a new login with SGID button, which sends the user to the SGID servers to start the login process
4. Add a new front end route, `/login/sgid/redirect`, which processes auth data from SGID + renders work email list for selection.

### PR Notes
This PR implements _most of_ step 3, except that the client ID is hard-coded to staging for now. I'll add in prod client ID once I change vite to support prod vs staging env vars in a separate PR.

Also note that this code relies on `crypto.randomUUID` - it should be well supported across our user base, but I added an explicit failure warning just in case there's some population group for which this doesn't work (e.g. people with funky virtual browsers).

#### Screenshots
**INITIAL STATE**
![Screenshot 2023-09-11 at 2 11 56 PM](https://github.com/opengovsg/plumber/assets/135598754/31cd7453-df43-433c-9472-afb6c3df4674)

**ERROR STATE**
![Screenshot 2023-09-11 at 2 09 34 PM](https://github.com/opengovsg/plumber/assets/135598754/f9125570-5ff8-4fc6-ade4-2629d2da337b)

## Tests
- Check that I get redirected to SGID login page after clicking the button
- Check that this is supported on GSIB machines.
- Regression test: check that I can still login via OTP.
